src/main/kotlin/org/rust/ide/actions/RustStringLiteralJoinLinesHandler.kt

```diff
@@ -0,0 +1,40 @@
+package org.rust.ide.actions
+
+import com.intellij.codeInsight.editorActions.JoinLinesHandlerDelegate
+import com.intellij.codeInsight.editorActions.JoinLinesHandlerDelegate.CANNOT_JOIN
+import com.intellij.openapi.editor.Document
+import com.intellij.psi.PsiFile
+import com.intellij.util.text.CharSequenceSubSequence
+import org.rust.ide.typing.endsWithUnescapedForwardSlash
+import org.rust.lang.core.psi.RustTokenElementTypes
+import org.rust.lang.core.psi.impl.RustFile
+import org.rust.lang.core.psi.util.elementType
+
+class RustStringLiteralJoinLinesHandler : JoinLinesHandlerDelegate {
+    override fun tryJoinLines(document: Document, file: PsiFile, offsetNear: Int, end: Int): Int {
+        if (file !is RustFile) return CANNOT_JOIN
+
+        val text = document.charsSequence
+        val leftPsi = file.findElementAt(offsetNear) ?: return CANNOT_JOIN
+        val rightPsi = file.findElementAt(end) ?: return CANNOT_JOIN
+
+        if (leftPsi == rightPsi && leftPsi.elementType in RustTokenElementTypes.STRING_LITERALS) {
+            var start = offsetNear
+
+            // Strip newline escape
+            if (CharSequenceSubSequence(text, 0, start + 1).endsWithUnescapedForwardSlash()) {
```

Comment:
Omg IntelliJ does string copying of entire document here o_O [source](https://upsource.jetbrains.com/idea-ce/file/idea-ce-5418ee4421e14e3eeb4cb10eca6407ee9618379e/plugins/properties/src/com/intellij/lang/properties/PropertiesJoinLinesHandler.java?nav=1162:1200:focused&line=0)